Implement `drop` argument for `[`, `subsetHz()`, `glom()`, `trunc()` SPC methods

brownag opened the pull request:
SPC `[` method and `subsetHz()` gain drop=TRUE argument. Also, harmonize with the `fill` argument used in `glom()` and make sure that arguments can be passed through from `trunc()`

  - `drop=FALSE` preserves one NA-filled placeholder horizon per profile(profile ID with all other values `NA`) where a horizon-subset would normally remove the profile
    - site data (other than the profile ID) also become `NA` with `drop=FALSE`
 
``` r
library(aqp)
#> This is aqp 2.0

data("jacobs2000")

jacobs2000[, 6:8]
#> SoilProfileCollection with 6 profiles and 11 horizons
#> profile ID: id  |  horizon ID: hzID 
#> Depth range: 152 - 213 cm
#> 
#> ----- Horizons (6 / 11 rows  |  10 / 18 columns) -----
#>    id hzID top bottom name sand clay matrix_color time_saturated
#>  92-1    6 153    156  B/C   41   50         <NA>              0
#>  92-1    7 156    213    C   50   47    #984C27FF              0
#>  92-2   13 145    213    C   49   40    #984C27FF              9
#>  92-3   19 112    165 Btv2   71   20    #AE9069FF             66
#>  92-3   20 165    175 Btv3   85   13    #BEAD98FF             78
#>  92-4   26 165    185  Cg2   96    1    #D6C9AEFF             41
#>  matrix_color_munsell
#>                  <NA>
#>             2.5YR 4/8
#>             2.5YR 4/8
#>              10YR 6/4
#>              10YR 7/2
#>              2.5Y 8/2
#> [... more horizons ...]
#> 
#> ----- Sites (6 / 6 rows  |  1 / 1 columns) -----
#>    id
#>  92-1
#>  92-2
#>  92-3
#>  92-4
#>  92-6
#>  92-7
#> 
#> Spatial Data:
#> [EMPTY]
jacobs2000[, 6:8, drop = FALSE]
#> SoilProfileCollection with 7 profiles and 12 horizons
#> profile ID: id  |  horizon ID: hzID 
#> Depth range: 152 - 213 cm
#> 
#> ----- Horizons (6 / 12 rows  |  10 / 18 columns) -----
#>    id hzID top bottom name sand clay matrix_color time_saturated
#>  92-1    6 153    156  B/C   41   50         <NA>              0
#>  92-1    7 156    213    C   50   47    #984C27FF              0
#>  92-2   13 145    213    C   49   40    #984C27FF              9
#>  92-3   19 112    165 Btv2   71   20    #AE9069FF             66
#>  92-3   20 165    175 Btv3   85   13    #BEAD98FF             78
#>  92-4   26 165    185  Cg2   96    1    #D6C9AEFF             41
#>  matrix_color_munsell
#>                  <NA>
#>             2.5YR 4/8
#>             2.5YR 4/8
#>              10YR 6/4
#>              10YR 7/2
#>              2.5Y 8/2
#> [... more horizons ...]
#> 
#> ----- Sites (6 / 7 rows  |  1 / 1 columns) -----
#>    id
#>  92-1
#>  92-2
#>  92-3
#>  92-4
#>  92-5
#>  92-6
#> [... more sites ...]
#> 
#> Spatial Data:
#> [EMPTY]

subsetHz(jacobs2000, grepl("2.5Y ", matrix_color_munsell))
#> SoilProfileCollection with 5 profiles and 8 horizons
#> profile ID: id  |  horizon ID: hzID 
#> Depth range: 43 - 185 cm
#> 
#> ----- Horizons (6 / 8 rows  |  10 / 18 columns) -----
#>    id hzID top bottom name sand clay matrix_color time_saturated
#>  92-1    2  18     43    E   80    6    #B1914AFF              0
#>  92-3   17  64     84   Bt   73   17    #B1914AFF             23
#>  92-4   26 165    185  Cg2   96    1    #D6C9AEFF             41
#>  92-5   29  28     61   E1   89    0    #C5AD7DFF              0
#>  92-5   30  61    109   E2   94    2    #C0AE89FF              0
#>  92-6   34  18     46   AE   82    2    #A59371FF              1
#>  matrix_color_munsell
#>              2.5Y 6/6
#>              2.5Y 6/6
#>              2.5Y 8/2
#>              2.5Y 7/4
#>              2.5Y 7/3
#>              2.5Y 6/3
#> [... more horizons ...]
#> 
#> ----- Sites (5 / 5 rows  |  1 / 1 columns) -----
#>    id
#>  92-1
#>  92-3
#>  92-4
#>  92-5
#>  92-6
#> 
#> Spatial Data:
#> [EMPTY]
subsetHz(jacobs2000, grepl("2.5Y ", matrix_color_munsell), drop = FALSE)
#> SoilProfileCollection with 7 profiles and 10 horizons
#> profile ID: id  |  horizon ID: hzID 
#> Depth range: 43 - 185 cm
#> 
#> ----- Horizons (6 / 10 rows  |  10 / 18 columns) -----
#>    id hzID top bottom name sand clay matrix_color time_saturated
#>  92-1    2  18     43    E   80    6    #B1914AFF              0
#>  92-2 <NA>  NA     NA <NA>   NA   NA         <NA>             NA
#>  92-3   17  64     84   Bt   73   17    #B1914AFF             23
#>  92-4   26 165    185  Cg2   96    1    #D6C9AEFF             41
#>  92-5   29  28     61   E1   89    0    #C5AD7DFF              0
#>  92-5   30  61    109   E2   94    2    #C0AE89FF              0
#>  matrix_color_munsell
#>              2.5Y 6/6
#>                  <NA>
#>              2.5Y 6/6
#>              2.5Y 8/2
#>              2.5Y 7/4
#>              2.5Y 7/3
#> [... more horizons ...]
#> 
#> ----- Sites (6 / 7 rows  |  1 / 1 columns) -----
#>    id
#>  92-1
#>  92-2
#>  92-3
#>  92-4
#>  92-5
#>  92-6
#> [... more sites ...]
#> 
#> Spatial Data:
#> [EMPTY]
```

dylanbeaudette commented:
This is a good idea, thanks for implementing it. I noticed something unexpected in the current master branch:
```r
jacobs2000[, 6:8]

# Error in validObject(.Object) : 
#invalid class “SoilProfileCollection” object: FALSE
```

Did implicit dropping of profiles break outside of this PR, or is that also implemented here?

brownag commented:
> Did implicit dropping of profiles break outside of this PR, or is that also implemented here?

I hadn't noticed it was broken until I went in to do this, but it is now fixed in this PR.

dylanbeaudette commented:
Great thanks. I like the general idea of maintaining vector order and length whenever possible.  There are a couple of other functions (`dice`, `slab`, ...?) that will drop profiles, but I'm still on the fence about those. Especially since they make such large modifications to the source data.

I'll review next. Initial testing with `plotSPC` makes me think that it will gracefully handle empty profiles across a large range of possible arguments--not something that I've previously tested. 

I'd like to have a fallback viz for empty profiles, like on SoilWeb. I'll implement that after this PR has been merged.

brownag commented:
> I'd like to have a fallback viz for empty profiles, like on SoilWeb. I'll implement that after this PR has been merged.

That sounds good! Will go a long way for interpreting SPCs with empty profiles

I think that w/ this PR (via new function: `insert_dropped_horizons()`)  the `drop=FALSE` (or "backfilling" of dropped profiles) can be implemented for any function that returns an SPC (or something resembling an SPCs site or horizon slots) that have the same profile IDs.

dylanbeaudette commented:
> I think that w/ this PR (via new function: `.insert_dropped_horizons()`) the `drop=FALSE` (or "backfilling" of dropped profiles) can be implemented for any function that returns an SPC (or something resembling an SPCs site or horizon slots) that has the same profile IDs.

Along those lines, do we need an `.isEmpty()` style function for unambiguously testing? Is there a simple way to use one of the example data sets + single function for dropping 1 or more sets of horizons for testing purposes? Something other than hz-subsetting?

dylanbeaudette commented on R/SoilProfileCollection-methods.R:
```diff
@@ -411,23 +414,72 @@ setMethod("subsetHz", signature(x = "SoilProfileCollection"), function(x, ...) {
+    replaceHorizons(x) <- newhz
+  } else {
+    # reinsert empty horizons and site data for "retained" profiles
+    x <- .insert_dropped_horizons(x, newhz)
```
OK, just to be sure: this replaces `x` entirely. This requires special attention to ensure the resulting SPC isn't altered in unexpected ways. Would it be simpler / safer to implement via `replaceHorizons()` on a modified `newhz`?

brownag replied:
There are two modes of usage of the internal function... one that returns a SPC (replaceHorizons() is used internally) and another that returns just a list of the data.frames from site and horizon.

brownag commented:
> Along those lines, do we need an `.isEmpty()` style function for unambiguously testing? 

I think a function like that would be good to export for easy checking for profiles where number of horizons ==1 and top and bottom depth are `NA`. Minimally something like this, but may also want to check other non-profile ID columns are `NA`.

``` r
library(aqp)
#> This is aqp 2.0
isEmpty <- function(x, ...) {
  res <- rep(FALSE, length(x))
  idx <- !apply(x[x[, , .NHZ] == 1, ][, , .TOP, .BOTTOM], 1, function(y) all(is.na(y)))
  res[idx] <- TRUE
  res
}
data("jacobs2000")
isEmpty(jacobs2000)
#> [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE
isEmpty(trunc(jacobs2000, 190, 200, drop = FALSE))
#> [1]  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE
```

Currently I don't need it for this PR, the determination of "needs to be filled" is by comparison of the original SPC against candidate horizons for replacement.

>Is there a simple way to use one of the example data sets + single function for dropping 1 or more sets of horizons for testing purposes? Something other than hz-subsetting?

Yes, as above, glom(), trunc() and subsetHz() can create situations where profiles that don't meet the depth/filter criteria are normally dropped
